File: data/research_features.lua
-- Derived feature register source data.
--
-- Instrument-backed features are expanded from `data/research_instruments.lua`
-- by `sim/research_features.lua`, so a construct can never exist as an
-- instrument construct without a registered feature (or vice versa). Only the
-- metadata that actually differs per instrument or per construct is authored
-- here; nothing is defaulted silently by the expander.
--
-- `extraction_commit` deliberately does not live here. The register declares
-- *which module and config* own an extraction; the commit that actually ran it
-- is recorded per feature version in the dataset manifest, where it is a fact
-- rather than a promise.

---@alias ResearchFeatureGrain
---| "tick"
---| "decision"
---| "possession"
---| "encounter"
---| "match"
---| "player_match"
---| "condition_block"
---| "session"
---| "participant"

---@alias ResearchFeatureLeakage
---| "none"
---| "outcome_derived"
---| "privileged_state"
---| "future_information"
---| "label_adjacent"

---@alias ResearchFeatureObservability
---| "player_observable"
---| "privileged_diagnostic"
---| "outcome_derived"
---| "protected_sensitive"
---| "prohibited"

---@alias ResearchEvidenceTier
---| "human_experience"
---| "behavioral_proxy"
---| "soccer_shape_proxy"
---| "machine_diagnostic"

---@alias ResearchOutcomeRole
---| "primary_outcome"
---| "secondary_outcome"
---| "guardrail"
---| "diagnostic"
---| "exploratory"
---| "feature_only"

---@alias ResearchMissingValueBehavior
---| "na_row"
---| "drop_row"
---| "zero_when_defined"
---| "protocol_failure"

---@alias ResearchPseudoReplicationGuard
---| "independent_unit_participant"
---| "cluster_by_participant"
---| "cluster_by_match"
---| "not_an_analysis_unit"

---@class ResearchCausalWindowData
---@field kind "same_tick"|"forward_ticks"|"backward_ticks"|"whole_match"|"post_condition"
---@field ticks integer?

---@class ResearchFeatureData
---@field id string
---@field version integer
---@field description string
---@field grain ResearchFeatureGrain
---@field source_schemas string[]
---@field source_fields string[]
---@field extraction_module string
---@field extraction_config_id string
---@field numerator string
---@field denominator string?
---@field unit string
---@field exclusions string[]
---@field missing_value_behavior ResearchMissingValueBehavior
---@field causal_window ResearchCausalWindowData
---@field normalization string
---@field leakage_risk ResearchFeatureLeakage
---@field observability ResearchFeatureObservability
---@field evidence_tier ResearchEvidenceTier
---@field outcome_role ResearchOutcomeRole
---@field aggregation_levels ResearchFeatureGrain[]
---@field pseudo_replication_guard ResearchPseudoReplicationGuard
---@field confounds string[]
---@field goodhart_failure string
---@field prohibited_uses string[]
---@field human_fun_claim boolean

---@class ResearchInstrumentFeatureDefaults
---@field grain ResearchFeatureGrain
---@field unit string
---@field extraction_module string
---@field extraction_config_id string
---@field source_schemas string[]
---@field exclusions string[]
---@field missing_value_behavior ResearchMissingValueBehavior
---@field causal_window ResearchCausalWindowData
---@field normalization string
---@field leakage_risk ResearchFeatureLeakage
---@field observability ResearchFeatureObservability
---@field evidence_tier ResearchEvidenceTier
---@field aggregation_levels ResearchFeatureGrain[]
---@field pseudo_replication_guard ResearchPseudoReplicationGuard
---@field prohibited_uses string[]

---@class ResearchConstructNoteData
---@field outcome_role ResearchOutcomeRole
---@field confounds string[]
---@field goodhart_failure string

---@class ResearchFeatureSourceData
---@field version integer
---@field instrument_defaults table<string, ResearchInstrumentFeatureDefaults>
---@field construct_notes table<string, ResearchConstructNoteData>
---@field behavioral ResearchFeatureData[]

---@type ResearchFeatureSourceData
return {
    version = 1,
    instrument_defaults = {
        pxi_enjoyment_addon = {
            grain = "condition_block",
            unit = "pxi_points_minus3_to_plus3",
            extraction_module = "sim.research_response",
            extraction_config_id = "pxi-enjoyment-mean-1",
            source_schemas = { "research_response_set/v1", "research_session_envelope/v1" },
            exclusions = { "practice_block", "excluded_session", "withdrawn_session" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "post_condition" },
            normalization = "none",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "human_experience",
            aggregation_levels = { "condition_block", "session", "participant" },
            pseudo_replication_guard = "independent_unit_participant",
            prohibited_uses = { "tick_row_analysis", "per_frame_weighting" },
        },
        pxi_partial_mechanisms = {
            grain = "condition_block",
            unit = "pxi_points_minus3_to_plus3",
            extraction_module = "sim.research_response",
            extraction_config_id = "pxi-subscale-mean-1",
            source_schemas = { "research_response_set/v1", "research_session_envelope/v1" },
            exclusions = { "practice_block", "excluded_session", "withdrawn_session" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "post_condition" },
            normalization = "none",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "human_experience",
            aggregation_levels = { "condition_block", "session", "participant" },
            pseudo_replication_guard = "independent_unit_participant",
            prohibited_uses = {
                "full_pxi_administration_claim",
                "benchmark_comparison",
                "tick_row_analysis",
            },
        },
        bangs_session = {
            grain = "condition_block",
            unit = "bangs_points_1_to_7",
            extraction_module = "sim.research_response",
            extraction_config_id = "bangs-subscale-mean-1",
            source_schemas = { "research_response_set/v1", "research_session_envelope/v1" },
            exclusions = { "practice_block", "excluded_session", "withdrawn_session" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "post_condition" },
            normalization = "none",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "human_experience",
            aggregation_levels = { "condition_block", "session", "participant" },
            pseudo_replication_guard = "independent_unit_participant",
            prohibited_uses = {
                "confirmatory_gate",
                "collapsing_frustration_into_satisfaction",
                "tick_row_analysis",
            },
        },
        affective_slider = {
            grain = "session",
            unit = "normalized_0_to_1",
            extraction_module = "sim.research_response",
            extraction_config_id = "affective-raw-1",
            source_schemas = { "research_response_set/v1", "research_session_envelope/v1" },
            exclusions = { "practice_block", "excluded_session", "withdrawn_session" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "post_condition" },
            normalization = "none",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "human_experience",
            aggregation_levels = { "session", "participant" },
            pseudo_replication_guard = "independent_unit_participant",
            prohibited_uses = {
                "pooling_with_custom_affect_fallback",
                "emotion_inference",
                "tick_row_analysis",
            },
        },
        custom_affect_fallback = {
            grain = "session",
            unit = "likert_points_1_to_7",
            extraction_module = "sim.research_response",
            extraction_config_id = "goliseo-affect-fallback-1",
            source_schemas = { "research_response_set/v1", "research_session_envelope/v1" },
            exclusions = { "practice_block", "excluded_session", "withdrawn_session" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "post_condition" },
            normalization = "none",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "human_experience",
            aggregation_levels = { "session", "participant" },
            pseudo_replication_guard = "independent_unit_participant",
            prohibited_uses = {
                "pooling_with_affective_slider",
                "validated_instrument_claim",
                "emotion_inference",
            },
        },
        custom_diagnostics = {
            grain = "condition_block",
            unit = "likert_points_minus3_to_plus3",
            extraction_module = "sim.research_response",
            extraction_config_id = "goliseo-custom-items-1",
            source_schemas = { "research_response_set/v1", "research_session_envelope/v1" },
            exclusions = { "practice_block", "excluded_session", "withdrawn_session" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "post_condition" },
            normalization = "none",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "human_experience",
            aggregation_levels = { "condition_block", "session", "participant" },
            pseudo_replication_guard = "independent_unit_participant",
            prohibited_uses = {
                "validated_scale_claim",
                "scale_score_aggregation",
                "confirmatory_gate",
            },
        },
    },
    construct_notes = {
        enjoyment = {
            outcome_role = "primary_outcome",
            confounds = { "novelty", "period_and_order", "win_or_loss", "fatigue" },
            goodhart_failure = "tuning for post-condition self-report while play quality falls",
        },
        autonomy = {
            outcome_role = "diagnostic",
            confounds = { "option_visibility", "role_assignment" },
            goodhart_failure = "adding options that raise perceived choice without agency",
        },
        mastery = {
            outcome_role = "diagnostic",
            confounds = { "learning_curve_position", "opponent_policy_strength" },
            goodhart_failure = "weakening opponents to manufacture felt mastery",
        },
        challenge = {
            outcome_role = "diagnostic",
            confounds = { "opponent_policy_strength", "experience_source_measures" },
            goodhart_failure = "treating mid-scale challenge as a target regardless of enjoyment",
        },
        ease_of_control = {
            outcome_role = "guardrail",
            confounds = { "control_device", "input_latency", "control_mapping" },
            goodhart_failure = "removing depth to raise ease scores",
        },
        goals_and_rules = {
            outcome_role = "guardrail",
            confounds = { "tutorial_exposure", "instruction_wording" },
            goodhart_failure = "over-explaining in instructions rather than in the game",
        },
        progress_feedback = {
            outcome_role = "diagnostic",
            confounds = { "score_state", "presentation_volume" },
            goodhart_failure = "adding feedback spectacle without informative content",
        },
        autonomy_satisfaction = {
            outcome_role = "exploratory",
            confounds = { "session_length", "condition_order" },
            goodhart_failure = "reading exploratory subscale movement as a confirmatory win",
        },
        autonomy_frustration = {
            outcome_role = "exploratory",
            confounds = { "session_length", "condition_order" },
            goodhart_failure = "collapsing frustration into a satisfaction composite",
        },
        competence_satisfaction = {
            outcome_role = "exploratory",
            confounds = { "win_or_loss", "opponent_policy_strength" },
            goodhart_failure = "inflating win rate to raise competence scores",
        },
        competence_frustration = {
            outcome_role = "exploratory",
            confounds = { "win_or_loss", "control_device" },
            goodhart_failure = "hiding failure feedback instead of making failure legible",
        },
        relatedness_satisfaction = {
            outcome_role = "exploratory",
            confounds = { "observer_mode", "facilitator_presence" },
            goodhart_failure = "reading facilitator rapport as game-driven relatedness",
        },
        relatedness_frustration = {
            outcome_role = "exploratory",
            confounds = { "observer_mode", "opponent_population" },
            goodhart_failure = "removing social framing rather than fixing the cause",
        },
        valence = {
            outcome_role = "diagnostic",
            confounds = { "break_timing", "instrument_modality" },
            goodhart_failure = "optimizing momentary affect at a natural break",
        },
        arousal = {
            outcome_role = "diagnostic",
            confounds = { "break_timing", "spectacle_volume" },
            goodhart_failure = "treating arousal as enjoyment",
        },
        soccer_primacy = {
            outcome_role = "guardrail",
            confounds = { "participant_genre_preference" },
            goodhart_failure = "suppressing combat entirely to protect the item",
        },
        fairness = {
            outcome_role = "guardrail",
            confounds = { "win_or_loss", "opponent_policy" },
            goodhart_failure = "removing counterplay variance to raise fairness ratings",
        },
        suspense = {
            outcome_role = "exploratory",
            confounds = { "score_closeness", "match_length" },
            goodhart_failure = "engineering close scores to manufacture suspense",
        },
        counterplay_readability = {
            outcome_role = "guardrail",
            confounds = { "cue_volume", "display_size", "readability_settings" },
            goodhart_failure = "adding cue noise that raises perceived readability",
        },
        overload = {
            outcome_role = "guardrail",
            confounds = { "render_performance", "session_position" },
            goodhart_failure = "cutting simultaneous options to lower overload",
        },
        frustration = {
            outcome_role = "guardrail",
            confounds = { "win_or_loss", "disable_duration" },
            goodhart_failure = "removing all punishment to lower frustration",
        },
        desire_to_explore = {
            outcome_role = "exploratory",
            confounds = { "novelty", "unrewarded_choice_framing" },
            goodhart_failure = "treating rematch clicks as intrinsic curiosity",
        },
    },
    behavioral = {
        {
            id = "soccer_shape_proxy_score",
            version = 1,
            description = "Geometric mean of banded MatchMetrics desirabilities. A soccer-shape proxy over simulated match statistics; it is not a measurement of human fun and carries no participant evidence.",
            grain = "match",
            source_schemas = { "gameplay_trace_manifest/v1" },
            source_fields = { "sim.metrics.MatchMetrics", "sim.metrics.bands" },
            extraction_module = "sim.metrics",
            extraction_config_id = "metrics-fun-score-bands-1",
            numerator = "product of per-band desirabilities present in the match",
            denominator = "count of banded metrics present, as a geometric-mean exponent",
            unit = "index_0_to_1",
            exclusions = { "matches with no banded metric present" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "whole_match" },
            normalization = "geometric_mean_over_present_bands",
            leakage_risk = "outcome_derived",
            observability = "privileged_diagnostic",
            evidence_tier = "soccer_shape_proxy",
            outcome_role = "diagnostic",
            aggregation_levels = { "match" },
            pseudo_replication_guard = "cluster_by_match",
            confounds = {
                "band midpoints are authored targets rather than measured preferences",
                "bot policy strength drives most banded metrics",
            },
            goodhart_failure = "tuning search maximizes band membership while human enjoyment falls",
            prohibited_uses = {
                "human_fun_claim",
                "primary_outcome",
                "proceed_gate",
                "participant_facing_report",
            },
            human_fun_claim = false,
        },
        {
            id = "involuntary_disable_share",
            version = 1,
            description = "Share of eligible active-play ticks in which a player was forcibly unable to act.",
            grain = "player_match",
            source_schemas = { "gameplay_trace_manifest/v1", "research_event_stream/v1" },
            source_fields = { "CombatPlayerState.forced_ticks", "MatchState.ball_in_play" },
            extraction_module = "sim.research_features",
            extraction_config_id = "combat-control-safety-1",
            numerator = "ticks with forced_ticks > 0",
            denominator = "eligible active-play ticks",
            unit = "share_0_to_1",
            exclusions = { "non-active-play ticks", "keeper slots", "practice_block" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "whole_match" },
            normalization = "share_of_denominator",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "behavioral_proxy",
            outcome_role = "guardrail",
            aggregation_levels = { "player_match", "match", "condition_block" },
            pseudo_replication_guard = "cluster_by_match",
            confounds = {
                "voluntary commitment mislabeled as disable",
                "disconnect or process exit truncating the denominator",
            },
            goodhart_failure = "shortening matches to lower the disable count",
            prohibited_uses = { "human_fun_claim", "participant_capability_inference" },
            human_fun_claim = false,
        },
        {
            id = "combat_to_soccer_conversion_rate",
            version = 1,
            description = "Share of confirmed accepted combat encounters that reach a soccer consequence inside the attribution window.",
            grain = "encounter",
            source_schemas = { "research_event_stream/v1" },
            source_fields = {
                "ResearchEventRow.domain",
                "ResearchEventRow.event_kind",
                "ResearchEventRow.source_sequence",
            },
            extraction_module = "sim.research_features",
            extraction_config_id = "combat-funnel-attribution-1",
            numerator = "confirmed encounters with an attributed soccer consequence",
            denominator = "confirmed accepted encounters",
            unit = "share_0_to_1",
            exclusions = {
                "rejected equipment requests",
                "revoked speculative encounters",
                "encounters whose window crosses full time",
            },
            missing_value_behavior = "na_row",
            causal_window = { kind = "forward_ticks", ticks = 120 },
            normalization = "share_of_denominator",
            leakage_risk = "label_adjacent",
            observability = "player_observable",
            evidence_tier = "behavioral_proxy",
            outcome_role = "diagnostic",
            aggregation_levels = { "encounter", "match", "condition_block" },
            pseudo_replication_guard = "cluster_by_match",
            confounds = {
                "attribution window length dominates the rate",
                "possession context differs by side and policy",
            },
            goodhart_failure = "widening the attribution window until every encounter looks causal",
            prohibited_uses = { "human_fun_claim", "speculative_event_inclusion" },
            human_fun_claim = false,
        },
        {
            id = "final_score_margin",
            version = 1,
            description = "Signed home-minus-away goal margin at full time. Outcome-derived: it is a label, not an input.",
            grain = "match",
            source_schemas = { "gameplay_trace_manifest/v1" },
            source_fields = { "sim.metrics.MatchMetrics.margin" },
            extraction_module = "sim.metrics",
            extraction_config_id = "metrics-final-margin-1",
            numerator = "home goals minus away goals at full time",
            unit = "goals",
            exclusions = { "incomplete matches", "practice_block" },
            missing_value_behavior = "drop_row",
            causal_window = { kind = "whole_match" },
            normalization = "none",
            leakage_risk = "outcome_derived",
            observability = "outcome_derived",
            evidence_tier = "behavioral_proxy",
            outcome_role = "diagnostic",
            aggregation_levels = { "match", "condition_block" },
            pseudo_replication_guard = "cluster_by_match",
            confounds = { "side assignment", "opponent policy strength" },
            goodhart_failure = "treating a wider margin as a better match experience",
            prohibited_uses = { "human_fun_claim", "model_input" },
            human_fun_claim = false,
        },
        {
            id = "declared_readability_settings",
            version = 1,
            description = "Readability settings the participant chose, recorded because they change what could be seen on screen.",
            grain = "session",
            source_schemas = { "research_session_envelope/v1" },
            source_fields = { "research_session_envelope.environment.readability_settings" },
            extraction_module = "sim.research_session",
            extraction_config_id = "session-readability-1",
            numerator = "the recorded readability setting map for the session",
            unit = "setting_map",
            exclusions = { "sessions with no recorded settings" },
            missing_value_behavior = "na_row",
            causal_window = { kind = "post_condition" },
            normalization = "none",
            leakage_risk = "none",
            observability = "protected_sensitive",
            evidence_tier = "machine_diagnostic",
            outcome_role = "feature_only",
            aggregation_levels = { "session" },
            pseudo_replication_guard = "cluster_by_participant",
            confounds = { "defaults left untouched are not a preference" },
            goodhart_failure = "reading a settings choice as a statement about the participant",
            prohibited_uses = {
                "human_fun_claim",
                "trait_inference",
                "disability_inference",
                "model_input",
                "per_participant_report",
            },
            human_fun_claim = false,
        },
        {
            id = "inferred_participant_skill_trait",
            version = 1,
            description = "Placeholder for an inferred per-participant skill or trait score. Registered so that referencing it fails the use gate instead of quietly appearing in a pipeline.",
            grain = "participant",
            source_schemas = { "research_session_envelope/v1", "gameplay_trace_manifest/v1" },
            source_fields = { "any combination of behavioral and session fields" },
            extraction_module = "none",
            extraction_config_id = "prohibited-no-extraction",
            numerator = "not computed; this feature has no implementation by design",
            unit = "prohibited",
            exclusions = { "every row: this feature is never extracted" },
            missing_value_behavior = "protocol_failure",
            causal_window = { kind = "whole_match" },
            normalization = "none",
            leakage_risk = "label_adjacent",
            observability = "prohibited",
            evidence_tier = "machine_diagnostic",
            outcome_role = "feature_only",
            aggregation_levels = { "participant" },
            pseudo_replication_guard = "not_an_analysis_unit",
            confounds = { "skill is confounded with device, role, and opponent policy" },
            goodhart_failure = "ranking testers by an invented skill number",
            prohibited_uses = {
                "any_analysis",
                "human_fun_claim",
                "model_input",
                "trait_inference",
                "per_participant_report",
            },
            human_fun_claim = false,
        },
        {
            id = "settled_turnovers_per_active_minute",
            version = 1,
            description = "Settled possession changes per active minute, as a soccer-integrity guardrail.",
            grain = "match",
            source_schemas = { "gameplay_trace_manifest/v1" },
            source_fields = { "sim.metrics.MatchMetrics.turnovers_per_min" },
            extraction_module = "sim.metrics",
            extraction_config_id = "metrics-settled-turnovers-1",
            numerator = "settled team possession changes",
            denominator = "active-play minutes",
            unit = "count_per_minute",
            exclusions = { "ownership flicker below the settle hold", "practice_block" },
            missing_value_behavior = "protocol_failure",
            causal_window = { kind = "whole_match" },
            normalization = "per_active_minute",
            leakage_risk = "none",
            observability = "player_observable",
            evidence_tier = "behavioral_proxy",
            outcome_role = "guardrail",
            aggregation_levels = { "match", "condition_block" },
            pseudo_replication_guard = "cluster_by_match",
            confounds = { "settle-hold threshold choice", "side and skill asymmetry" },
            goodhart_failure = "raising raw ownership churn and calling it dynamism",
            prohibited_uses = { "human_fun_claim" },
            human_fun_claim = false,
        },
    },
}

File: rust/crates/gc-sim/tests/research_features.rs
//! Tests for `gc_sim::research_features`.

use gc_sim::research_features;
use gc_sim::research_schema::{self, Value};

fn set_field(value: &Value, name: &str, new_value: Value) -> Value {
    let entries = value.as_record().expect("record value").to_vec();
    let mut updated: Vec<(String, Value)> =
        entries.into_iter().filter(|(key, _)| key != name).collect();
    updated.push((name.to_string(), new_value));
    Value::Record(updated)
}

fn field_str<'a>(value: &'a Value, name: &str) -> &'a str {
    let entries = value.as_record().expect("record value");
    Value::record_get(entries, name)
        .and_then(Value::as_str)
        .expect("field present")
}

fn field_bool(value: &Value, name: &str) -> bool {
    let entries = value.as_record().expect("record value");
    Value::record_get(entries, name)
        .and_then(Value::as_bool)
        .expect("field present")
}

#[test]
fn research_feature_register_accepts_the_authored_register() {
    assert!(research_features::validate_registry());
    assert!(research_features::ids().len() > 20);
}

#[test]
fn research_feature_register_registers_every_instrument_construct_exactly_once() {
    let mut expected = 0;
    for instrument in gc_data::research_instruments::ALL {
        for &construct in instrument.constructs {
            expected += 1;
            let id = research_features::feature_id(instrument.id, construct);
            let feature = research_features::feature(&id).expect("registered feature");
            assert_eq!(field_str(&feature, "evidence_tier"), "human_experience");
        }
    }
    assert_eq!(expected, 24, "the instrument register changed shape");
    assert_eq!(research_features::ids().len(), expected + 7);
}

#[test]
fn research_feature_register_returns_copies_so_callers_cannot_edit_the_register() {
    let feature = research_features::feature("soccer_shape_proxy_score").unwrap();
    let _mutated = set_field(&feature, "human_fun_claim", Value::Bool(true));
    let refetched = research_features::feature("soccer_shape_proxy_score").unwrap();
    assert!(!field_bool(&refetched, "human_fun_claim"));
}

#[test]
fn fun_score_tagging_registers_metrics_fun_score_as_a_soccer_shape_proxy_never_human_fun() {
    let feature = research_features::feature("soccer_shape_proxy_score").unwrap();
    assert_eq!(field_str(&feature, "evidence_tier"), "soccer_shape_proxy");
    assert!(!field_bool(&feature, "human_fun_claim"));
    assert_eq!(field_str(&feature, "outcome_role"), "diagnostic");
    assert_eq!(field_str(&feature, "leakage_risk"), "outcome_derived");
    assert_eq!(
        field_str(&feature, "observability"),
        "privileged_diagnostic"
    );
    assert!(field_str(&feature, "description").contains("not a measurement of human fun"));

    let err =
        research_features::use_allowed("soccer_shape_proxy_score", "human_fun_claim").unwrap_err();
    assert!(err.contains("prohibits use"), "{err}");
    assert!(research_features::use_allowed("soccer_shape_proxy_score", "primary_outcome").is_err());
    assert!(research_features::use_allowed("soccer_shape_proxy_score", "model_input").is_ok());
}

#[test]
fn fun_score_tagging_keeps_the_human_fun_claim_on_the_primary_enjoyment_endpoint_only() {
    let mut claims = Vec::new();
    for id in research_features::ids() {
        let feature = research_features::feature(&id).unwrap();
        if field_bool(&feature, "human_fun_claim") {
            claims.push(id);
        }
    }
    assert_eq!(claims.len(), 1);
    assert_eq!(claims[0], "pxi_enjoyment_addon.enjoyment");

    let enjoyment = research_features::feature("pxi_enjoyment_addon.enjoyment").unwrap();
    assert_eq!(field_str(&enjoyment, "outcome_role"), "primary_outcome");
    assert_eq!(field_str(&enjoyment, "leakage_risk"), "none");
    assert_eq!(field_str(&enjoyment, "evidence_tier"), "human_experience");
}

#[test]
fn fun_score_tagging_prohibits_promoting_a_diagnostic_or_exploratory_construct() {
    for id in [
        "pxi_partial_mechanisms.autonomy",
        "bangs_session.competence_frustration",
        "custom_diagnostics.fairness",
        "affective_slider.valence",
    ] {
        let err = research_features::use_allowed(id, "primary_outcome");
        assert!(err.is_err(), "{id} must prohibit primary_outcome use");
    }
}

#[test]
fn research_feature_observability_categories_has_a_worked_example_for_every_observability_category()
{
    let mut seen: Vec<String> = Vec::new();
    for id in research_features::ids() {
        let feature = research_features::feature(&id).unwrap();
        let observability = field_str(&feature, "observability").to_string();
        if !seen.contains(&observability) {
            seen.push(observability);
        }
    }
    for category in [
        "player_observable",
        "privileged_diagnostic",
        "outcome_derived",
        "protected_sensitive",
        "prohibited",
    ] {
        assert!(
            seen.iter().any(|s| s == category),
            "no registered feature exercises {category}"
        );
    }
}

#[test]
fn research_feature_observability_categories_keeps_prohibited_and_sensitive_features_out_of_every_use_it_names()
 {
    let prohibited = research_features::feature("inferred_participant_skill_trait").unwrap();
    assert_eq!(field_str(&prohibited, "observability"), "prohibited");
    assert_eq!(field_str(&prohibited, "outcome_role"), "feature_only");
    assert!(
        research_features::use_allowed("inferred_participant_skill_trait", "any_analysis").is_err()
    );
    assert!(
        research_features::use_allowed("inferred_participant_skill_trait", "model_input").is_err()
    );

    let sensitive = research_features::feature("declared_readability_settings").unwrap();
    assert_eq!(
        field_str(&sensitive, "observability"),
        "protected_sensitive"
    );
    assert!(
        research_features::use_allowed("declared_readability_settings", "disability_inference")
            .is_err()
    );

    let outcome = research_features::feature("final_score_margin").unwrap();
    assert_eq!(field_str(&outcome, "observability"), "outcome_derived");
    assert_eq!(field_str(&outcome, "leakage_risk"), "outcome_derived");
    assert!(research_features::use_allowed("final_score_margin", "model_input").is_err());
}

#[test]
fn research_feature_grains_never_treats_a_tick_match_or_encounter_row_as_a_participant() {
    for id in research_features::ids() {
        let feature = research_features::feature(&id).unwrap();
        let grain = field_str(&feature, "grain");
        if research_features::is_clustered_grain(grain) {
            assert_ne!(
                field_str(&feature, "pseudo_replication_guard"),
                "independent_unit_participant",
                "{id} claims participant independence at grain {grain}"
            );
        }
    }
}

#[test]
fn research_feature_grains_only_allows_the_aggregation_levels_it_defines() {
    assert!(
        research_features::aggregation_allowed("pxi_enjoyment_addon.enjoyment", "condition_block")
            .is_ok()
    );
    let err = research_features::aggregation_allowed("pxi_enjoyment_addon.enjoyment", "tick")
        .unwrap_err();
    assert!(err.contains("aggregation level"), "{err}");
    assert!(research_features::aggregation_allowed("soccer_shape_proxy_score", "session").is_err());
    assert!(research_features::aggregation_allowed("no_such_feature", "match").is_err());
}

#[test]
fn research_feature_grains_rejects_a_feature_that_violates_the_register_invariants() {
    let base = research_features::feature("involuntary_disable_share").unwrap();
    assert!(research_features::validate_feature(&base, None).is_ok());

    let pseudo_replicating = set_field(
        &base,
        "pseudo_replication_guard",
        Value::str("independent_unit_participant"),
    );
    let err = research_features::validate_feature(&pseudo_replicating, None).unwrap_err();
    assert!(err.contains("independent participants"), "{err}");

    let proxy_claiming_fun = set_field(
        &research_features::feature("soccer_shape_proxy_score").unwrap(),
        "human_fun_claim",
        Value::Bool(true),
    );
    assert!(research_features::validate_feature(&proxy_claiming_fun, None).is_err());

    let leaky_primary = set_field(&base, "outcome_role", Value::str("primary_outcome"));
    assert!(research_features::validate_feature(&leaky_primary, None).is_err());

    let missing_window = set_field(
        &research_features::feature("combat_to_soccer_conversion_rate").unwrap(),
        "causal_window",
        Value::Record(vec![("kind".to_string(), Value::str("forward_ticks"))]),
    );
    assert!(research_features::validate_feature(&missing_window, None).is_err());

    let unregistered_grain = set_field(
        &base,
        "aggregation_levels",
        Value::Array(vec![Value::str("session")]),
    );
    assert!(research_features::validate_feature(&unregistered_grain, None).is_err());
}

#[test]
fn research_feature_register_hash_is_stable_and_covers_every_feature_definition() {
    let hash = research_features::registry_hash();
    assert_eq!(hash.len(), research_schema::HASH_LENGTH);
    assert_eq!(hash, research_features::registry_hash());

    let feature = research_features::feature("involuntary_disable_share").unwrap();
    let before = research_schema::content_hash(&research_features::shape(), &feature).unwrap();
    let changed = set_field(
        &feature,
        "causal_window",
        Value::Record(vec![
            ("kind".to_string(), Value::str("forward_ticks")),
            ("ticks".to_string(), Value::Number(60.0)),
        ]),
    );
    let after = research_schema::content_hash(&research_features::shape(), &changed).unwrap();
    assert_ne!(before, after);
}
